MAINT: 1.15.3 backports

tylerjereddy opened the pull request:
There are a few nasty bugs that have cropped up in the `1.15.x` series for which simple patches have been provided, so the release may be worth doing.

Backports Included here so far:

1. gh-22561
2. gh-22567
3. gh-22587
4. gh-22588
5. gh-22660
6. gh-22693
7. gh-22822
8. gh-22828
9. gh-22834
10. Partial backport of gh-22214 just to get `xp_default_dtype`, which was used in one of the above backports
11. gh-22810
12. gh-22690
13. gh-22792
14. gh-22784
15. gh-22772
16. gh-22763
17. gh-22689
18. gh-22643
19. gh-22642
20. gh-22600
21. gh-22582
22. gh-22555
23. gh-22532
24. gh-22502
25. gh-22517
26. gh-22726
27. gh-22617
28. gh-22590
29. gh-22869
30. gh-22910
31. gh-22608
32. gh-20035

----

**I considered for backporting but probably too many conflicts to pull in:**

1. gh-22554
2. gh-22624
3. gh-22760
4. gh-22802 (Cython->C transition happened in between?)

----

TODO:

- [x] Investigate cause of local test failures on this branch: `ImportError: cannot import name 'xp_default_dtype' from 'scipy._lib._array_api'`
- [x] manual combing through additional bug fixes that could be backported (there are usually several that slip through the cracks)
- [x] wrap up some open PRs with backport label (currently they are both waiting on me): https://github.com/scipy/scipy/pull/20035 and https://github.com/scipy/scipy/pull/22608
- [x] update release notes
- [x] gradually reactivate the CI, up to testing wheel builds 
- [x] final read-through of diff

ev-br committed:
ENH: add xp_default_dtype helper

[ci skip] [skip ci]

inkydragon committed:
BUG: special: Fix typo in specfun::chgu (#22502)

xref: https://github.com/scipy/scipy/blob/9d08eb58205fb669ff72c60080b18be61b14d8e1/scipy/special/specfun/specfun.f#L4705-L4715

nbelakovski committed:
CI: Use actions/cache 4.2.0

4.1.2 is deprecated, see https://github.com/scipy/scipy/issues/22515

nascheme committed:
BUG: _lib: Fix data race found by TSAN, use SCIPY_TLS.

The 'lowlevelcallable_type' static variable should be thread-local to
avoid races for free-threaded Python.  Closes gh-22547.

fbourgey committed:
MAINT: stats.genextreme.stats: don't warn when `c=0` (#22532)

MAINT: stats.genextreme.state: don't warn when `c=0`

---------

Co-authored-by: Matt Haberland <mhaberla@calpoly.edu>

steppi committed:
MAINT: optimize: fix `bracket_root` termination check and default initial bracket. (#22567)

* TST: Add test for failing bracket_root case
* MAINT: Fix edge case for bracket_root
* MAINT: Fix default for xr0 in bracket_root
* MAINT: Fix dtypes in default endpoints

---------

Co-authored-by: Matt Haberland <mhaberla@calpoly.edu>

rgommers committed:
BENCH: remove triple run of mmread/mmwrite benchmark, limit sizes

This fixes a problem on CircleCI. Closes gh-22574

This also reduced the runtime of the `io_mm` benchmarks by 10x,
it was super slow for no good reason because the benchmarks were
run 3 times, 2x with fast-matrix-market and 1x with the old pure
Python implementation. This looked like a left-over from the PR
that switched to fast-matrix-market, demonstrating that `io.mmread`
and `io.mmwrite` are the fast version and `io._mmio` is way slower.
There is no need to keep running that.

steppi committed:
BUG: Pin jupyterlite-sphinx to >= 0.19.1 (#22587)

Romain-Geissler-1A committed:
BLD: Force defining the mdspan parenthesis operator even with C++ >= 23. (#22588)

This allows to keep the same code, independently of the compiler C++
dialect being used. The default behavior of kokkos mdspan is the
following:
 - if __cpp_multidimensional_subscript is defined (ie C++ >= 23), then
   enable by default the square bracket operator to access elements, and
   disable by default the parenthesis operator.
 - otherwise (C++ < 23), disable the square bracket operator which
   cannot work, and replace it by the parenthesis operator instead.

The current xsf code always uses the parenthesis operator. To avoid
having to always write the two possibilities, force the kokkos mdspan
implementation to always define the parenthesis operator for now.

Signed-off-by: Romain Geissler <romain.geissler@amadeus.com>

sarthakg004 committed:
BUG: Fix ArpackError default argument (#22600)

czgdp1807 committed:
CI: Windows: Use only intel Fortran essentials to minimise space usage

czgdp1807 committed:
CI: Windows: Download intel oneAPI before setting up conda environment

czgdp1807 committed:
CI: Linux: Update Intel oneAPI URLs to use only Fortran and C++ components

czgdp1807 committed:
CI: Linux: Download Intel oneAPI before setting up conda environment

czgdp1807 committed:
BUG: `integrate.solve_ivp`: Avoid duplicate time steps in `ts` when `dense_output=True` (#22582)

* DEV: Append to ts and ys only if their length is greater than 1 and t != ts[-1]

* TEST: Added regression test

* MISC: Styling changes

* TEST: Improve test as per reviews

dschult committed:
enforce 1D sum/mean out parameter correct shape (#22642)

hubernikus committed:
BUG: spatial.transform.Rotation: support 0-length rotations (#22643)

* Enable zero-dimensional cython.view.array

* Explicit size in quat-compose.

* Remove 0 quat exception.

* Exception for mean of 0-Rotation.

* Test for zero-length array.

* Remove 0-size input exception.

* Change RuntimeWarning to ValueError on mean of 0-lenght Rotation.

* Change test to value error.

* Error out slerp for 0-rotatoin + test.

* Additional 0-lenght Rotation tests.

* Improve exception message of mean of empty rotation.

* Test boolean indexing of rotation.

* Fix concatentation test.

* String matching for value error in test.

* Fix power in test-loop.

* Separate test function and additional assert statements.

* Rename message variable

* Remove duplicate assert statement.

* Remove duplicate assert.

* Remove space.

* One liner for quat-compose-shape.

* Remove hanging space.

* Fix linter errors.

nickodell committed:
MAINT: optimize._highs_wrapper: Avoid looking up basis.col_status each time (#22660)

tylerjereddy committed:
BUG: resample_poly dtype preservation

* Fixes gh-22684.

* Fix a scenario where `resample_poly` did not preserve input
`dtype`.

rsbowen committed:
BUG: spatial.HalfspaceIntersection: fix `add_halfspaces` batch size support (#22693)

* Add a (failing) test that checks we can add halfspaces with various batch sizes
* Fix dims for dists in the case that there is more than one halfspace being added
* Add correctness, not just doesnt-error, test
* Make the feasible point not the origin so that the dist array is not all the same
* Update the docs of add_halfspaces to look more like the constructor, to
make it crystal clear what the shapes should be, as well as the fact
that these new half spaces need to contain the original feasible point.


---------

Co-authored-by: Tyler Reddy <tyler.je.reddy@gmail.com>

tylerjereddy committed:
MAINT: compensate for dot exceptions

* Fixes gh-22720.

* This patch adds a few floating point exception
handling shims to compensate for the upstream
change for `np.dot` at https://github.com/numpy/numpy/pull/28442,
merged five days ago. This allows the Scipy testsuite to pass
again locally on x86_64 Linux against NumPy `main`.

czgdp1807 committed:
BUG: scipy.sparse.linalg: Fix expm_multiply if both A and B are sparse (#22555)

---------
Co-authored-by: Jake Bowhay <jb9.bowhay@gmail.com>

DVarela56 committed:
BUG/TST:sparse.linalg: Transposed LinearOperator multiplication

Fixes #22637
Added a case in `LinearOperator._rmatvec` to use
`_rmatmat` as a fallback when self._adjoint is not defined
Included a test that ensures that LinearOperator handles
transposed multiplications when only _rmatmat is defined

mdhaber committed:
MAINT/DOC: integrate.tanhsinh: lightly refactor error estimate and correct documentation (#22690)

* MAINT/DOC: integrate.tanhsinh: lightly refactor error estimate for simplicity and correct documentation

* TST: integrate.tanhsinh: add unit test for NaN error issue

* MAINT: special.tanhsinh: implement minimal change to fix gh-22681

DietBru committed:
BUG: signal.ShortTimeFFT.upper_border_begin: Document parameter `n`. (#22792)

In `ShortTimeFFT.upper_border_begin` the parameter `n` is not documented, which is fixed here. Furthermore:
* Fix the same problem in `ShortTimeFFT._post_padding` and `ShortTimeFFT.k_max`
* Document return parameters in `ShortTimeFFT.upper_border_begin`.
* Test that `n` is large enough in `ShortTimeFFT.upper_border_begin` and `ShortTimeFFT._post_padding` (including unit test).
* Fix small variable name inconsistenc in `ShortTimeFFT` docstr.

nickodell committed:
BUG: sparse.linalg._AdjointLinearOperator: Remove reference cycle to improve memory use (#22763)

* BUG: sparse: fix reference cycle in MatrixLinearOperator

* ENH: sparse: remove duplicate variable

* TST: sparse: add test for MLO refcycle

sildater committed:
Update test_short_time_fft.py upper_border_begin

Due to PR #22792 n needs to be ≥ len(w)/2.

Co-authored-by: Dietrich Brunn <12721170+DietBru@users.noreply.github.com>

ilayn committed:
BUG: linalg.expm: Fix noncompliant compiler branch typos in C code (#22822)

ilayn committed:
BUG: linalg: Fix shape mismatch in interpolative.svd

ilayn committed:
MAINT: linalg: Remove the leftover F77 pyf file from id_dist

ilayn committed:
TST: linalg: Add missing empty line at the end in test_interpolative

[lint only]

lucascolley committed:
Merge pull request #22828 from lucascolley/22236

BUG: add workaround for pytest assertion rewriting overreach

rgommers commented:
Thanks Tyler. What's in here so far looks good to me, and one more bug fix release can't hurt indeed given that we have 3 months to go to the next feature release (assuming making CI happy isn't too much trouble).

lucascolley commented:
> Investigate cause of local test failures on this branch

https://github.com/scipy/scipy/pull/22840/commits/83d07bdfac16fdb5c46ed91328e956190c58af73 may require pulling in https://github.com/scipy/scipy/commit/6199db65a684e9b2773c18cb9000c4df1d2c7401

czgdp1807 committed:
BUG: special.hyp2f1: fix for extreme inputs (#22810)

* Regression testing and patch for extreme input handling in `hyp2f1`. Fixes gh-20988.

tylerjereddy commented:
Thanks Lucas, that allows the full suite to pass locally on this branch on ARM Mac. I pushed it up. Not ready to flush CI yet here though.

It might be argued that array API infra improvements shouldn't really be backported, but I think practicality over purity on this one for now given how small that patch is. Can always revisit when I check the final diff here.

tylerjereddy committed:
TST: PR 22840 revisions

* Clean up a botched backport of gh-22784 (some
testing shims were missing).

[ci skip] [skip ci]

tylerjereddy commented:
After manually combing through bug fix PRs, the number of backport PRs included here has roughly tripled to 24, and the full testsuite still passes locally (after some wrangling with merge conflicts). Something weird hapened when cherry-picking gh-22784; long story short, merging multiple different branches into a PR branch seems like a bad idea for backport-targeting PRs but I sorted it out.

I made a small list of additional PRs I considered for backporting but for which the merge conflicts were just too wild to sensibly deal with (in the updated original comment above).

tylerjereddy committed:
DOC: update 1.15.3 relnotes

* Update the SciPy `1.15.3` release notes following
substantial backport activity.

tylerjereddy commented:
Visual inspection of the updated/rendered release notes locally seems "ok," proceeding with regular CI testing for a bit.

I see that an issue has already cropped up: 

> This request has been automatically failed because it uses a deprecated version of `actions/cache: 6849a6489940f00c2f30c0fb92c6274307ccb58a``

Anyway, the point is to discover issues in the regular CI that require additional backport activity (this is almost always needed after more than few weeks have passed since a previous release).

ilayn commented:
Ah yes that NNLS one is based on a feature coming in 1.16. I completely missed that. My apologies for the noise. Hence it can't be backported before 1.16. You can safely ignore that one.

tylerjereddy commented:
Regular CI is getting a bit healthier now, remaining issues there include:

- `asv` failure in circleci
- Intel oneAPI CI failure re: `Not enough disk space.`
- Linux CI failures with `RuntimeWarning: invalid value encountered in dot`; need to backport my own patch from https://github.com/scipy/scipy/pull/22726
- non-default interpreter job failure with `TypeError: canonicalize_name() got an unexpected keyword argument 'validate'`

smurfix committed:
Merge pull request #22869 from smurfix/main

BUG: optimize._highspy: don't import from inside a C module

tylerjereddy committed:
CI: PR 22840 revisions

* Update the version of `packaging` used for the non-default
interpreter CI job for the reasons described at:
https://github.com/scipy/scipy/pull/22840#issuecomment-2817494211

tylerjereddy committed:
CI, MAINT: PR 22840 revisions

* The array API CI job was failing because `torch`
was missing at test run time. Try to match the "actions"
versions used on `main` branch, since there aren't many
other differences in the CI configs on these branches.

tylerjereddy commented on .github/workflows/array_api.yml:
```diff
@@ -95,7 +95,7 @@ jobs:
 
     - name: Setup compiler cache
-      uses:  actions/cache@6849a6489940f00c2f30c0fb92c6274307ccb58a # v4.1.2
+      uses:  actions/cache@5a3ec84eff668545956fd18022155c47e93e2684 # v4.2.3
```
There aren't many relevant differences with `main` for this array API CI config, so I was hoping these bumps might help with the issue `NameError: name 'torch' is not defined`.

But they didn't, and the full array API-activated test suite passes locally on CPU on this branch.

tylerjereddy commented on .github/workflows/linux.yml:
```diff
@@ -428,8 +428,11 @@ jobs:
+        # https://github.com/scipy/scipy/pull/22840#issuecomment-2817494211
         run: |
           python3.11 -m pip install build pythran pybind11 cython numpy meson-python
+          python3.11 -m pip install -U packaging
```
This is annoying shim to add--the justification is at https://github.com/scipy/scipy/pull/22840#issuecomment-2817494211 -- something may have changed in the runner image to allow an older/problematic `packaging` to be picked up from a local/cached version.

Feels a bit hacky..

rgommers replied:
That is indeed annoying. Surprising it doesn't happen on `main`. Either way, feels like a metadata issue in meson-python, I'll have a look.

rgommers replied:
Opened https://github.com/mesonbuild/meson-python/issues/733

tylerjereddy committed:
MAINT, CI: PR 22840 revisions

* The `torch` version used in the array API CI job
was causing problems:
https://github.com/scipy/scipy/pull/22840#discussion_r2057109473

* Pin `torch` to version `2.6.0` since this was previously
passing tests on the maintenance branch and `2.7.0` was released
very recently (April 23/2025, yesterday at the time of writing)

tylerjereddy commented:
The `torch` import error causing the remaining CI failures was reproducible locally when upgrading to `torch==2.7.0` on this feature branch and then running the array API testsuite; i.e., `python dev.py test -t scipy/fft/tests/test_basic.py::TestFFT::test_all_1d_norm_preserving -b torch` will suffice to repro in that env.

`torch` `2.7.0` was released April 23 (yesterday!) so I'm feeling less bad about a breakage with that--I'm not sure that we have a guaranteed support window for array API backends given that this all still experimental, at least on this branch. If we did start adopting support windows for the major array API backend deps that might be a bit of a project if they have dev cycles that deviate substantially from i.e., the NumPy N+3 business.

We could probably bisect this easily enough to find the cause of the "break" in `torch`, which presumably has been shimmed around either in array-api-compat or our own internal array API shims (both have substantial diffs I think). We could also just document the practical upper bound in the relnotes given these observations I suppose.

I've tentatively pinned to `torch==2.6.0` to see if that gets to all-green regular CI for now.

tylerjereddy commented:
Ok, all green now. I won't proceed to wheel build testing yet--a few more backports on the radar and I'll let the hacks used here settle a little bit.

mdhaber commented:
@tylerjereddy Do you have a moment to test the MRE in gh-22903 with this branch? I finally got the boost subpackage to find the commit in `main`, but now when I try to go back to this branch, it is having trouble finding 260f7600e95dcf87b228d46a4a37cf594df5437b.

mdhaber commented:
After that, it still can't find the commit.
```
(scipy-dev) matthaberland@Matts-MacBook-Air-6 scipy % git submodule update --init
fatal: transport 'file' not allowed
fatal: Fetched in submodule path 'scipy/_lib/boost_math', but it did not contain 260f7600e95dcf87b228d46a4a37cf594df5437b. Direct fetching of that commit failed
```

I had a similar error last night when I was trying to update submodules after getting the latest main. The usual tricks weren't working. I finally found way for it to find the commit it needs for `main`, but now it can't find the commit it needs for this branch.

lucascolley commented:
One difference between `main` and this branch is that the submodule has moved to `subprojects/boost_math/math` on `main`. That shouldn't affect things if you are starting from this branch though. Maybe inspect your `.git/modules` for anything that seems mismatched with the branch you are trying to replicate locally.

tylerjereddy commented:
> @tylerjereddy Do you have a moment to test the MRE in https://github.com/scipy/scipy/issues/22903 with this branch?

@mdhaber on the latest version of this branch I see `nan` for the provided reproducer on local M3 mac. On same machine on latest `main` I see `1.245216515790739`.

Let me know if you want me to backport something else, or apply the label to it. I need to make a push to pull in a final set of backports pretty "soon."

mdhaber commented:
I don't know which of these commands does the trick but after:
```
git fetch --unshallow
git fetch origin --depth=1000000
git fetch origin '+refs/heads/*:refs/remotes/origin/*' --update-head-ok
```
it is able to find the commit. These same commands fixed the problems I had when we made the switch to Boost as a subproject.

mdhaber committed:
MAINT: special.logsumexp: fix bug when weight of largest magnitude component is negative

tylerjereddy committed:
BUG: spatial.HalfspaceIntersection: raise on non-feasible half space (#20035)

* Fixes #19865

* Rather than returning a problematic half space intersection
polytope, we raise an error for incremental processing when
a new half space is added for which the interior point is no
longer clearly interior

---------

Co-authored-by: Richard Strong Bowen <richard@geomagical.com>

aiudirog committed:
BUG: ndimage.median_filter: fix segfault when using `mode='mirror'` (#22608)

* It appears there was an off-by-one error in the loop that finalizes the mirror mode.

* Fix another memory issue in `_rank_filter()` based on feedback
at https://github.com/scipy/scipy/pull/22608#discussion_r2067771870.
This seems to allow many repeats of the property-based test to finally
pass on this branch via i.e.,
`spin test -t scipy/ndimage/tests/test_filters.py::test_gh_22586_crash_property -- --count=50`

---------

Co-authored-by: Tyler Reddy <tyler.je.reddy@gmail.com>

tylerjereddy committed:
DOC: PR 22840 revisions

* Update the release notes following additional
backport activity for SciPy `1.15.3`.

tylerjereddy committed:
MAINT: PR 22840 revisions

* Remove an unused import highlighted by the linter.

tylerjereddy commented:
Updates:

- this backport PR is now at parity with the merged backport PR list (32 merged, 0 open)
- full test suite passes locally on latest version of the branch; relnotes updated and still render "ok" locally after doc build
- if the regular CI passes here I'll proceed with: 1) skimming through the diff/comments for any remaining issues; 2) probing wheel builds

tylerjereddy commented:
array API fails probably need fixing

tylerjereddy committed:
MAINT: PR 22840 revisions

* `logsumexp()` had some array API compatibility issues
on the release branch related to the version of the array
API standard and possibly other dependency versions that
are older than on `main` branch. This patch works around
these by guarding `xp.real()` to only apply to complex
input and coercing a complex Python scalar to an array
to deal with a promotion issue:
`TypeError: Python complex scalars can only be promoted with complex floating-point arrays.`

tylerjereddy commented:
The array API failures (sample log: https://github.com/scipy/scipy/actions/runs/14847812524/job/41685521405?pr=22840) have been worked around with a minor patch in `_logsumexp.py` that compensates for some type promotion issues and the capability expansion for `xp.real()` that happened in a more recent version of the standard not in use on the release branch.

The patch I applied ([bd0f132](https://github.com/scipy/scipy/pull/22840/commits/bd0f132b37f5eaab79b50eefbe2139c0f1bde443)) is pasted below the fold for convenience

<details>

```diff
--- a/scipy/special/_logsumexp.py
+++ b/scipy/special/_logsumexp.py
@@ -7,6 +7,7 @@ from scipy._lib._array_api import (
     xp_broadcast_promote,
     xp_copy,
     xp_float_to_complex,
+    is_complex,
 )
 from scipy._lib import array_api_extra as xpx
 
@@ -232,13 +233,14 @@ def _logsumexp(a, b, axis, return_sign, xp):
         m = xp.abs(m)
     else:
         # `a_max` can have a sign component for complex input
-        sgn = sgn * xp.exp(xp.imag(a_max) * 1.0j)
+        sgn = sgn * xp.exp(xp.imag(a_max) * xp.asarray(1.0j, dtype=a_max.dtype))
 
     # Take log and undo shift
     out = xp.log1p(s) + xp.log(m) + a_max
 
     if return_sign:
-        out = xp.real(out)
+        if is_complex(out, xp):
+            out = xp.real(out)
     elif xp.isdtype(out.dtype, 'real floating'):
         out[sgn < 0] = xp.nan
```

</details>

@mdhaber @crusaderky this part of the code has seen a very large number of array API-related mutations on `main` and release branch support for the array API remains experimental. Nonetheless, the full tests do seem to pass for array API locally with that patch. If the CI agrees here and you don't object I'll probably try to move forward with the next steps in the release process.

tylerjereddy committed:
MAINT: PR 22840 wheel build [wheel build]

* An empty commit to trigger a test wheel build
prior to the SciPy `1.15.3` release process.

[wheel build]

tylerjereddy commented:
With regular CI passing, I re-read the full diff and was reasonably happy with it (recognized ~90 % of the changes, didn't see anything suspicious). None of the remaining review comments/discussions are blocking as far as I can tell.

I've initiated a test wheel build to see where we stand on the binaries--if/when those are all green I think we should move forward, but let's see what the wheel builds decide.

tylerjereddy commented:
Wheel builds all succeeded, the only CI failure is an irrelevant "fail slow" fluke on Windows.

As such, I'll start the `1.15.3` release process but probably won't finish it until tomorrow (May 8th).

crusaderky commented:
> The array API failures (sample log: https://github.com/scipy/scipy/actions/runs/14847812524/job/41685521405?pr=22840) have been worked around with a minor patch in `_logsumexp.py` that compensates for some type promotion issues and the capability expansion for `xp.real()` that happened in a more recent version of the standard not in use on the release branch.
> 
> The patch I applied ([bd0f132](https://github.com/scipy/scipy/pull/22840/commits/bd0f132b37f5eaab79b50eefbe2139c0f1bde443)) is pasted below the fold for convenience
> 
> ```diff
> --- a/scipy/special/_logsumexp.py
> +++ b/scipy/special/_logsumexp.py
> @@ -7,6 +7,7 @@ from scipy._lib._array_api import (
>      xp_broadcast_promote,
>      xp_copy,
>      xp_float_to_complex,
> +    is_complex,
>  )
>  from scipy._lib import array_api_extra as xpx
>  
> @@ -232,13 +233,14 @@ def _logsumexp(a, b, axis, return_sign, xp):
>          m = xp.abs(m)
>      else:
>          # `a_max` can have a sign component for complex input
> -        sgn = sgn * xp.exp(xp.imag(a_max) * 1.0j)
> +        sgn = sgn * xp.exp(xp.imag(a_max) * xp.asarray(1.0j, dtype=a_max.dtype))
>  
>      # Take log and undo shift
>      out = xp.log1p(s) + xp.log(m) + a_max
>  
>      if return_sign:
> -        out = xp.real(out)
> +        if is_complex(out, xp):
> +            out = xp.real(out)
>      elif xp.isdtype(out.dtype, 'real floating'):
>          out[sgn < 0] = xp.nan
> ```
> 
> @mdhaber @crusaderky this part of the code has seen a very large number of array API-related mutations on `main` and release branch support for the array API remains experimental. Nonetheless, the full tests do seem to pass for array API locally with that patch. If the CI agrees here and you don't object I'll probably try to move forward with the next steps in the release process.

It's a CI artefact.
Your CI pinned array-api-strict 2.1.0, which is doesn't offer Array API 2024.12 compliance.
The issue disappears when upgrading to array-api-strict 2.3.1.

tylerjereddy commented:
It looks like you added support for `2024.12` well after the `1.15.x` series in https://github.com/scipy/scipy/pull/22687, so I don't think we have any obligation to support that here, especially since it remains experimental.

Ideally backports would be tailored to match the array API standard state of the release branch unless we start adopting a policy of supporting a wider set of standards, but that seems impractical on release branches for now.
